<?xml version="1.0" encoding="utf-8"?>

<robot name="robot_description">

<material name="red">
  <color rgba="1.0 0.0 0.0 1.0"/>
</material>

<!-- %%%%%%%%%%%%%%%%%%%%%%%%%%%% LINKS %%%%%%%%%%%%%%%%%%%%%%%%%%%% -->
  <!-- %%%%%%%%%%%%%%%%%%%%%%% BASE LINK %%%%%%%%%%%%%%%%%%%%%%% -->
    <link name="base_link">
      <inertial>
        <origin
          xyz="-1.5853E-16 5.2101E-17 -0.015636"
          rpy="0 0 0" />
        <mass
          value="0.36521" />
        <inertia
          ixx="0.00039748"
          ixy="-4.0792E-09"
          ixz="7.1512E-18"
          iyy="0.00072454"
          iyz="-1.3047E-18"
          izz="0.0010802" />
      </inertial>
      <visual>
        <origin
          xyz="0 0 0"
          rpy="0 0 0" />
        <geometry>
          <mesh
            filename="package://robot_description/meshes/base_link.STL" />
        </geometry>
        <material
          name="">
          <color
            rgba="0.75294 0.75294 0.75294 1" />
        </material>
      </visual>
      <collision>
        <origin
          xyz="0 0 0"
          rpy="0 0 0" />
        <geometry>
          <mesh
            filename="package://robot_description/meshes/base_link.STL" />
        </geometry>
      </collision>
    </link>
  <!-- %%%%%%%%%%%%%%%%%%%%% FR WHEEL LINK %%%%%%%%%%%%%%%%%%%%% -->
    <link name="fr_wheel_link">
      <inertial>
        <origin
          xyz="1.8041E-16 -0.011571 5.9328E-16"
          rpy="0 0 0" />
        <mass
          value="0.041912" />
        <inertia
          ixx="3.6789E-06"
          ixy="7.9409E-23"
          ixz="-2.3641E-21"
          iyy="3.5671E-06"
          iyz="8.8839E-22"
          izz="3.6789E-06" />
      </inertial>
      <visual>
        <origin
          xyz="0 0 0"
          rpy="1.5708 0 0" />
        <geometry>
          <cylinder radius="0.048" length="0.02" />
        </geometry>
        <material
          name="">
          <color
            rgba="0.75294 0.75294 0.75294 1" />
        </material>
      </visual>
      <collision>
        <origin
          xyz="0 0 0"
          rpy="1.5708 0 0" />
        <geometry>
          <cylinder radius="0.048" length="0.02" />
        </geometry>
      </collision>   
    </link>
  <!-- %%%%%%%%%%%%%%%%%%%%% FL WHEEL LINK %%%%%%%%%%%%%%%%%%%%% -->
    <link name="fl_wheel_link">
      <inertial>
        <origin
          xyz="5.4956E-15 0.011595 -7.8063E-16"
          rpy="0 0 0" />
        <mass
          value="0.041912" />
        <inertia
          ixx="3.6789E-06"
          ixy="-1.1316E-21"
          ixz="9.3084E-21"
          iyy="3.5669E-06"
          iyz="-6.9925E-22"
          izz="3.6789E-06" />
      </inertial>
      <visual>
        <origin
          xyz="0 0 0"
          rpy="1.5708 0 0" />
        <geometry>
          <cylinder radius="0.048" length="0.02" />
        </geometry>
        <material
          name="">
          <color
            rgba="0.75294 0.75294 0.75294 1" />
        </material>
      </visual>
      <collision>
        <origin
          xyz="0 0 0"
          rpy="1.5708 0 0" />
        <geometry>
          <cylinder radius="0.048" length="0.02" />
        </geometry>
      </collision>
    </link>
  <!-- %%%%%%%%%%%%%%%%%%%%% BR WHEEL LINK %%%%%%%%%%%%%%%%%%%%% -->
    <link name="br_wheel_link">
      <inertial>
        <origin
          xyz="5.7454E-15 -0.011499 -2.8103E-16"
          rpy="0 0 0" />
        <mass
          value="0.041912" />
        <inertia
          ixx="3.6789E-06"
          ixy="-5.9722E-22"
          ixz="9.4729E-21"
          iyy="3.5669E-06"
          iyz="-2.4485E-22"
          izz="3.6789E-06" />
      </inertial>
      <visual>
        <origin
          xyz="0 0 0"
          rpy="1.5708 0 0" />
        <geometry>
          <cylinder radius="0.048" length="0.02" />
        </geometry>
        <material
          name="">
          <color
            rgba="0.75294 0.75294 0.75294 1" />
        </material>
      </visual>
      <collision>
        <origin
          xyz="0 0 0"
          rpy="1.5708 0 0" />
        <geometry>
          <cylinder radius="0.048" length="0.02" />
        </geometry>
      </collision>
    </link>
  <!-- %%%%%%%%%%%%%%%%%%%%% BL WHEEL LINK %%%%%%%%%%%%%%%%%%%%% -->
    <link name="bl_wheel_link">
      <inertial>
        <origin
          xyz="1.1102E-16 0.011429 -1.6653E-16"
          rpy="0 0 0" />
        <mass
          value="0.041912" />
        <inertia
          ixx="3.6789E-06"
          ixy="-1.5032E-22"
          ixz="-2.9973E-21"
          iyy="3.5671E-06"
          iyz="1.1701E-21"
          izz="3.6789E-06" />
      </inertial>
      <visual>
        <origin
          xyz="0 0 0"
          rpy="1.5708 0 0" />
        <geometry>
          <cylinder radius="0.048" length="0.02" />
        </geometry>
        <material
          name="">
          <color
            rgba="0.75294 0.75294 0.75294 1" />
        </material>
      </visual>
      <collision>
        <origin
          xyz="0 0 0"
          rpy="1.5708 0 0" />
        <geometry>
          <cylinder radius="0.048" length="0.02" />
        </geometry>
      </collision>
    </link>

<!-- %%%%%%%%%%%%%%%%%%%%%%%%%%%% JOINTS %%%%%%%%%%%%%%%%%%%%%%%%%%%% -->
  <!-- %%%%%%%%%%%%%%%%%%%% FR WHEEL JOINT %%%%%%%%%%%%%%%%%%%% -->
    <joint name="fr_wheel_joint"
      type="fixed">
      <origin
        xyz="0.10615 -0.085325 -0.0303"
        rpy="0 0 0" />
      <parent
        link="base_link" />
      <child
        link="fr_wheel_link" />
      <axis
        xyz="0 1 0" />
    </joint>
  <!-- %%%%%%%%%%%%%%%%%%%%% FL WHEEL JOINT %%%%%%%%%%%%%%%%%%%%% -->
    <joint name="fl_wheel_joint"
      type="fixed">
      <origin
        xyz="0.10615 0.085325 -0.0303"
        rpy="0 0 0" />
      <parent
        link="base_link" />
      <child
        link="fl_wheel_link" />
      <axis
        xyz="0 1 0" />
    </joint>
  <!-- %%%%%%%%%%%%%%%%%%%%% BR WHEEL JOINT %%%%%%%%%%%%%%%%%%%%% -->
    <joint name="br_wheel_joint"
      type="fixed">
      <origin
        xyz="-0.10615 -0.085325 -0.0303"
        rpy="0 0 0" />
      <parent
        link="base_link" />
      <child
        link="br_wheel_link" />
      <axis
        xyz="0 1 0" />
    </joint>
  <!-- %%%%%%%%%%%%%%%%%%%%% BR WHEEL JOINT %%%%%%%%%%%%%%%%%%%%% -->
    <joint name="bl_wheel_joint"
      type="fixed">
      <origin
        xyz="-0.10615 0.085325 -0.0303"
        rpy="0 0 0" />
      <parent
        link="base_link" />
      <child
        link="bl_wheel_link" />
      <axis
        xyz="0 1 0" />
    </joint>

    <joint name="laser_joint"
      type="fixed">
      <origin
        xyz="0.14 0 0.001"
        rpy="0 0 0" />
      <parent
        link="base_link" />
      <child
        link="laser_frame" />
    </joint>

    <link name="laser_frame">
      <visual>
        <geometry>
          <cylinder radius="0.02" length="0.02"/>
        </geometry>
        <material name="red">
          
        </material>
      </visual>

      <collision>
        <geometry>
          <cylinder radius="0.02" length="0.02"/>
        </geometry>
      </collision>

      <inertial>
        <origin
          xyz="0 0 0"
          rpy="0 0 0" />
        <mass
          value="0.01" />
        <inertia
          ixx="3.6789E-06"
          ixy="-1.5032E-22"
          ixz="-2.9973E-21"
          iyy="3.5671E-06"
          iyz="1.1701E-21"
          izz="3.6789E-06" />
      </inertial>
    </link>

  <gazebo reference="laser_frame">
    <material>Gazebo/Red</material>

    <sensor name="laser" type="ray">
        <pose> 0 0 0 0 0 0 </pose>
        <visualize>false</visualize>
        <update_rate>10</update_rate>
        <ray>
            <scan>
                <horizontal>
                    <samples>360</samples>
                    <min_angle>-3.14</min_angle>
                    <max_angle>3.14</max_angle>
                </horizontal>
            </scan>
            <range>
                <min>0.3</min>
                <max>12</max>
            </range>
        </ray>
        <plugin name="laser_controller" filename="libgazebo_ros_laser.so">
            <topicName>/scan</topicName>
            <frameName>laser_frame</frameName>
        </plugin>
    </sensor>
  </gazebo>

  <gazebo>
    <plugin name="object_controller" filename="libgazebo_ros_planar_move.so">
      <commandTopic>cmd_vel</commandTopic>
      <odometryTopic>odom</odometryTopic>
      <odometryFrame>odom</odometryFrame>
      <odometryRate>100.0</odometryRate>
      <robotBaseFrame>base_link</robotBaseFrame>
      <publishTF>true</publishTF>
    </plugin>
  </gazebo>

</robot>